cdap-securestore-ext-cloudkms/src/main/java/co/cask/cdap/securestore/spi/CloudSecretManager.java

```diff
@@ -0,0 +1,98 @@
+import com.google.common.collect.ImmutableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link SecretManager} which manages sensitive data using Google Cloud KMS.
+ */
+public class CloudSecretManager implements SecretManager {
+  private static final Logger LOG = LoggerFactory.getLogger(CloudSecretManager.class);
+  private static final String CLOUD_KMS_NAME = "cloudkms";
+  private SecretStore store;
+  private CloudKMSClient client;
+  private SecretEncoderDecoder encoderDecoder;
+
+  @Override
+  public String getName() {
+    return CLOUD_KMS_NAME;
+  }
+
+  @Override
+  public void initialize(SecretManagerContext context) throws IOException {
+    this.store = context.getSecretStore();
+    this.client = new CloudKMSClient();
+    this.client.createKeyRing();
+    this.encoderDecoder = new SecretEncoderDecoder();
+  }
+
+  @Override
+  public void store(String namespace, Secret secret) throws IOException {
+    // creates new crypto key for every namespace
+    client.createCryptoKey(namespace);
+    byte[] encryptedData = client.encrypt(namespace, secret.getData());
+    if (encryptedData == null) {
+      // this should not happen
+      throw new IOException(String.format("Could not encrypt secret %s. Encrypted secret is null.",
+                                          secret.getMetadata().getName()));
+    }
+    Secret encryptedSecret = new Secret(encryptedData, secret.getMetadata());
+    store.store(namespace, secret.getMetadata().getName(), encoderDecoder, encryptedSecret);
+  }
+
+  @Override
+  public Secret get(String namespace, String name) throws SecretNotFoundException, IOException {
+    Secret secret = store.get(namespace, name, encoderDecoder);
+    byte[] decrypted = client.decrypt(namespace, secret.getData());
+    if (decrypted == null) {
+      throw new IOException(String.format("Could not decrypt secret %s. Decrypted secret is null.",
+                                          secret.getMetadata().getName()));
+    }
+    return new Secret(decrypted, secret.getMetadata());
+  }
+
+  @Override
+  public Collection<SecretMetadata> list(String namespace) throws IOException {
+    Collection<Secret> secrets = store.list(namespace, encoderDecoder);
+    return secrets.stream().map(Secret::getMetadata).collect(Collectors.toCollection(ImmutableList::of));
+  }
+
+  @Override
+  public void delete(String namespace, String name) throws SecretNotFoundException, IOException {
+    // Cloud KMS does not provide capability to delete crypto keys.
+    // Reference: https://cloud.google.com/kms/docs/faq#cannnot_delete
+    store.delete(namespace, name);
+  }
+
+  @Override
+  public void destroy(SecretManagerContext context) {
+    try {
+      client.close();
+    } catch (IOException e) {
+      LOG.error("Error while closing cloud kms client.", e);
```

Comment:
Is it actually an Error? Meaning something will not work after this point? If not, log it with lower log level.